src/test/java/com/google/opensesame/projects/ProjectEntityTest.java

```diff
@@ -0,0 +1,162 @@
+import com.googlecode.objectify.ObjectifyFactory;
+import com.googlecode.objectify.ObjectifyService;
+import com.googlecode.objectify.util.Closeable;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.Arrays;
+import java.util.Collection;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class ProjectEntityTest {
+  private final LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
+  private Closeable session;
+
+  private ProjectEntity firstMockProject;
+  private ProjectEntity secondMockProject;
+  private ProjectEntity thirdMockProject;
+
+  private HttpServletResponse mockResponse;
+  private StringWriter responseStringWriter;
+
+  @BeforeClass
+  public static void objectifyClassSetup() {
+    ObjectifyService.setFactory(new ObjectifyFactory());
+    ObjectifyService.register(ProjectEntity.class);
+  }
+
+  @Before
+  public void objectifyStartSession() throws IOException {
+    session = ObjectifyService.begin();
+    helper.setUp();
+  }
+
+  @Before
+  public void mockProjectSetUp() {
+    firstMockProject =
+        new ProjectEntity(
+            "273537467", Arrays.asList("Mentor 1", "Mentor 2", "Mentor 3"), Arrays.asList());
+    secondMockProject =
+        new ProjectEntity("45717250", Arrays.asList("Mentor 1", "Mentor 2"), Arrays.asList());
+    thirdMockProject = new ProjectEntity("20580498", Arrays.asList("Mentor 1"), Arrays.asList());
+    ofy().save().entities(firstMockProject, secondMockProject, thirdMockProject).now();
+  }
+
+  @Before
+  public void mockResponseSetUp() throws IOException {
+    mockResponse = mock(HttpServletResponse.class);
+    responseStringWriter = new StringWriter();
+    when(mockResponse.getWriter()).thenReturn(new PrintWriter(responseStringWriter));
+  }
+
+  @After
+  public void objectifyEndSession() {
+    session.close();
+    helper.tearDown();
+  }
+
+  @Test
+  public void queryByIdsShouldReturnProjectsWithThoseIds() throws IOException {
+    // Expect to receive the ProjectEntities with the corresponding repository IDs.
+
+    HttpServletRequest request = mock(HttpServletRequest.class);
+    when(request.getParameterValues(ProjectEntity.PROJECT_ID_PARAM))
+        .thenReturn(new String[] {firstMockProject.repositoryId, secondMockProject.repositoryId});
+
+    Collection<ProjectEntity> queryResult = ProjectEntity.queryFromRequest(request, mockResponse);
+    Collection<ProjectEntity> expected = Arrays.asList(firstMockProject, secondMockProject);
+
+    assertNotNull(queryResult);
+    assertTrue(queryResult.size() == expected.size() && queryResult.containsAll(expected));
+    verify(mockResponse, never()).setStatus(anyInt());
```

Comment:
Seeing this code here makes me wonder if all our code is in the right place? IMO the *Entity classes are for interacting with Datastore and shouldn't have servlet concepts like Requests and Responses in them. It seems like the responsibilities are a little mixed. 

The fact that we have to check that `setStatus` is called instead of looking at a Response returned is a sign of that, to me. Not a huge deal, something I think might be worth refactoring in the future. The servlet should handle the request parsing and responses and all of that, and just pass the IDs to the Entity class.

Reply:
I completely agree. I'm thinking we could refactor this by introducing something like the [optional result](https://github.com/Richie78321/google-step/blob/master/portfolio/src/main/java/com/google/sps/helper/ValidationResult.java) that I introduced in my portfolio all those weeks ago. The main reason that I had given the responsibility of error responses to the individual functions was so that context wouldn't be lost from outside of the function if an error occurred, but with some sort of wrapper class we could preserve all of that context! I'll be adding this to the list of issues on the alpha project.